lib/events/writer.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package events
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"time"
+
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+)
+
+// NewWriterLog returns a new instance of writer log
+func NewWriterLog(w io.WriteCloser) *WriterLog {
+	return &WriterLog{
+		w:      w,
+		clock:  clockwork.NewRealClock(),
+		newUID: utils.NewRealUID(),
+	}
+}
+
+// WriterLog is an audit log that emits all events
+// to the external writer
+type WriterLog struct {
+	w     io.WriteCloser
+	clock clockwork.Clock
+	// newUID is used to generate unique IDs for events
+	newUID utils.UID
+}
+
+// Close releases connection and resources associated with log if any
+func (w *WriterLog) Close() error {
+	return w.w.Close()
+}
+
+// EmitAuditEvent emits audit event
+func (w *WriterLog) EmitAuditEvent(event Event, fields EventFields) error {
+	err := UpdateEventFields(event, fields, w.clock, w.newUID)
+	if err != nil {
+		log.Error(err)
+		// even in case of error, prefer to log incomplete event
+		// rather than to log nothing
+	}
+	// line is the text to be logged
+	line, err := json.Marshal(fields)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	_, err = w.w.Write(line)
+	return trace.ConvertSystemError(err)
+}
+
+// DELETE IN: 2.7.0
```

Comment:
Can this function be removed?

Reply:
I looked and it seems like some parts are still using it.

Reply:
Then let's just remove the DELETE IN line since it doesn't look like it's going to be removed?